src/ctapipe/io/interpolation.py

```diff
@@ -0,0 +1,283 @@
+from abc import ABCMeta, abstractmethod
+from typing import Any
+
+import astropy.units as u
+import numpy as np
+import tables
+from astropy.time import Time
+from scipy.interpolate import interp1d
+
+from ctapipe.core import Component, traits
+
+from .astropy_helpers import read_table
+
+
+class StepFunction:
+
+    """
+    Step function Interpolator for the gain and pedestals
+
+    Parameters
+    ----------
+    values : None | np.array
+        Numpy array of the data that is to be interpolated.
+        The first dimension needs to be an index over time
+    times : None | np.array
+        Time values over which data are to be interpolated
+        need to be sorted and have same length as first dimension of values
+    """
+
+    def __init__(
+        self,
+        times,
+        values,
+        bounds_error=True,
+        fill_value="extrapolate",
+        assume_sorted=True,
+        copy=False,
+    ):
+        self.values = values
+        self.times = times
+        self.bounds_error = bounds_error
+        self.fill_value = fill_value
+
+    def __call__(self, point):
+        if point < self.times[0]:
+            if self.bounds_error:
+                raise ValueError("below the interpolation range")
+
+            if self.fill_value == "extrapolate":
+                return self.values[0]
+
+            else:
+                a = np.empty(self.values[0].shape)
+                a[:] = np.nan
+                return a
+
+        else:
+            i = np.searchsorted(self.times, point, side="left")
+            return self.values[i - 1]
+
+
+class Interpolator(Component, metaclass=ABCMeta):
+    """
+    Interpolator parent class.
+
+    Parameters
+    ----------
+    h5file : None | tables.File
+        A open hdf5 file with read access.
+    """
+
+    bounds_error = traits.Bool(
+        default_value=True,
+        help="If true, raises an exception when trying to extrapolate out of the given table",
+    ).tag(config=True)
+
+    extrapolate = traits.Bool(
+        help="If bounds_error is False, this flag will specify whether values outside"
+        "the available values are filled with nan (False) or extrapolated (True).",
+        default_value=False,
+    ).tag(config=True)
+
+    table_location = None
```

Comment:
Should add a docstring or comment that these *class variables* should be set by sub-classes. You could instead _use instance variables_ and  require them to be set in the `__init__` instead of using class variables, and then the sub-classes would look like:

```py3
class SomeOtherInterpolator(Interpolator):
    def __init__(self):
         super().__init__(self, table_location="/dl0/something/table")
         ...
```
Not sure which is more clear, but using class variables is less common.   @maxnoe do you have a guideline here on when we should use class variables for defaults vs instance?

Reply:
I think this should be a class variable and not changeable by instance.

E.g. the path in the h5file is fixed for `PointingInterpolator`, `PedestalInterpolator`, whatever we come up with.

Reply:
It's a question of when this data changes, in the design I had in mind, each of the different interpolators looks to a fixed table path → class variable.

Reply:
Ok.  Then just have to fix the `add_table` method to not step on this